Articles/Blog/WhenIsBetterToWriteAutomatedTests.md

```diff
@@ -0,0 +1,133 @@
+And now after having had followed this road from 2007 until now having written and maintained thousands of automated tests over that time, I have come to realize that there are times when it is actually better not to write automated tests for some pieces of software!
+
+## Criteria for when it is better to not add automated tests
+
+I have come to realize that it often does not pay off to write automated tests for a piece of software when the following five things are true about that software:
+
+* **The damage done by a defect is minor.**
+* **The majority of defects will be obvious.**
+* **Fixing the majority of defects will be easy.**
+* **Manually testing the software after a change is easy.**
+* **Writing automated tests is hard or the tests will be hard to maintain.**
+
+If all five of the above criteria are satisfied, it is often better not to write automated tests for a piece of software.
+The time spent writing, maintaining, and running the automated tests does not have enough benefit to outweigh the cost.
+
+For example, some bash scripts used locally that load some modules and then configure, build, and run tests for some software may not be worth writing an automated test suite for because:
+
+* The damage done by not having everything complete correctly will be minor because you can just run it again.
+* A failure will be obvious because the software will not even build and no tests will be run.
+* The scripts do pretty basic stuff and will be easy to fix if they fail.
+* It would be difficult to write mock up the various commands run by the script.
+
+Therefore, it is often not worth the investment to write automated tests for locally run scripts of this type.
+
+## Example where automated tests should have been added
+
+However, if such scripts are being used to automate some important process like deploying the software to users where not performing the task correctly would cause non-trivial harm, then one should write some automated tests to protect the critical functionality of the scripts.
+Or, if other people are running the scripts and it will not be obvious to them that a failure has occurred or they will not know how to fix it quickly, then one should likely write some automated tests for the scripts.
+
+For example, there was a bash test-driver script in a CSE project where the script (run as a cron job) failed to detect that the code's tests were failing and instead sent out emails that everything was passing.
+As a result, the code was broken for weeks with no one noticing.
+A release of the software went out (to hundreds of internal users) and it was the users who noticed the new defects.
+This wasted user's time, may have resulted in incorrect results, and damaged the reputation of the project releasing the software.
+The lesson is that some *scripts* may actually need to be considered *software* in their own right and need to have strong automated testing just like any piece of non-trivial software.
```

Comment:
Most instead of "some"

Reply:
> Most instead of "some"

Not sure we have enough data for sure to say "most".  But no one can argue with "some".